[WIP] [ENH] Add ability to reject epochs using callables

withmywoessner opened the pull request:
#### Reference issue
Example: Fixes ##11775.


#### What does this implement/fix?
Add ability to reject epochs using functions.
Ex: epochs = mne.Epochs(..., reject=dict(eeg=lambda x: True if np.median(x) > 1e-5 else False))

withmywoessner commented:
I am still working on this. Does anyone know of any good testing data that has massive amplitude spikes?

larsoner commented:
The `testing` dataset should have some spikes due to blinks but I'm not sure they'd qualify as massive. You can always create one using `RawArray` or `EpochsArray`, and it can end up being nice and readable/explicit that way.

withmywoessner commented:
Thanks @larsoner! I am getting ready to write a tutorial (and the test though this question applies to both) as per @drammock's suggestion. First, I wanted to ask is the way I generated sample bad data appropriate or should I use the source simulator class (I did not really know how to get the desired large amplitude spike easily)? Here is what I made: 
```
data_path = mne.datasets.testing.data_path(download=False)
fname_raw_testing = data_path / "MEG" / "sample" / "sample_audvis_trunc_raw.fif"
raw = mne.io.read_raw_fif(fname_raw_testing)
raw.crop(0, 5)
raw.del_proj()
chans = raw.info['ch_names'][-6:-1]
raw.pick(chans)
data = raw.get_data()

new_data = data

# Make large spike
new_data[0, 180:200] *= 1e7

# Shift amp up to calculate median/mean later
new_data[0, 610:880] += 1e-3
edit_raw = mne.io.RawArray(new_data, raw.info)
```
Second, I wanted to include a section of the tutorial for including boolean operations such as:
```
epochs = mne.Epochs(edit_raw, 
                    events, 
                    tmin=0, 
                    tmax=1, 
                    baseline=None, 
                    reject=dict(eeg=lambda x: True if ((np.max(x, axis=1) > 1).any() or 
                                                       (np.median(x, axis=1) > 1e-3).any()) else False),
                    preload=True, verbose='DEBUG'
) 
```
In the tutorial is it better to place the complicated lambda function in the construction or declare a new function like:
```
# Preparing a function for the 'reject' parameter
def reject_criteria(x):
    max_condition = np.max(x, axis=1) > 1
    median_condition = np.median(x, axis=1) > 1e-3
    return True if max_condition.any() or median_condition.any() else False

# Creating epochs with specified parameters
epochs = mne.Epochs(
    raw=edit_raw, 
    events=events, 
    tmin=0, 
    tmax=1, 
    baseline=None, 
    reject=dict(eeg=reject_criteria),
    preload=True, 
    verbose='DEBUG'
)
```
Third, I was wondering if instead of modifying `reject `and `flat`, it would be better to just create a new parameter `reject_func` as it may be confusing to a user if `reject `or `flat` can do the same thing depending whether you use `>` or `<`.

Thanks for the help! :)

larsoner commented:
> First, I wanted to ask is the way I generated sample bad data appropriate

I think using SourceSimulator would be overkill and direct data modification is fine!

> In the tutorial is it better to place the complicated lambda function in the construction or declare a new function like:

New function is better/clearer, probably added as a short new section in https://mne.tools/stable/auto_tutorials/preprocessing/20_rejecting_bad_data.html / https://github.com/mne-tools/mne-python/blob/main/tutorials/preprocessing/20_rejecting_bad_data.py

> Third, I was wondering if instead of modifying reject and flat, it would be better to just create a new parameter reject_func as it may be confusing to a user if reject or flat can do the same thing depending whether you use > or <.

Good question. I think as long as we clarify in the doc that the callable could detect all sorts of stuff -- either too large or too small amplitudes -- and that what goes in `drop_log` reflects whether it came from `reject` or `flat` then it's fine.

withmywoessner commented:
Okay, I was just asking because I personally don't think I would ever use both if I was trying to specify a minimum and a maximum absolute value and would just use boolean operators. In the tutorial, is it fine if I use `reject` and booleans to specify the min/max or should I use both to show it is possible @larsoner?

withmywoessner commented:
> I think as long as we clarify in the doc that the callable could detect all sorts of stuff -- either too large or too small amplitudes -- and that what goes in `drop_log` ref

I was looking over the documentation and I see the drop_log as 4 main reasons listed:
<img width="421" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/157cca61-05cf-424c-b098-e136c67a58c6">
Currently, when you drop based on amplitude threshold (and my implementation) it just gives the channel name:
<img width="221" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/562f8e98-8c6a-45fd-a7b5-45e9daeb6e06">
Can I add a new reason, `'REJECT_CRITERIA'`, for when you drop based on `reject` or `flat` that includes the function or value used to drop? 

Also, should I keep the channel that had the bad item or just include the function?

withmywoessner commented on mne/epochs.py:
```diff
@@ -818,9 +818,14 @@ def _reject_setup(self, reject, flat):
+                        % (kind, val)
                     )
 
         # now check to see if our rejection and flat are getting more
```
# now check to see if our rejection and flat are getting more
        # restrictive
        old_reject = self.reject if self.reject is not None else dict()
        old_flat = self.flat if self.flat is not None else dict()
        bad_msg = (
            '{kind}["{key}"] == {new} {op} {old} (old value), new '
            "{kind} values must be at least as stringent as "
            "previous ones"
        )

        # copy thresholds for channel types that were used previously, but not
        # passed this time
        for key in set(old_reject) - set(reject):
            reject[key] = old_reject[key]
        # make sure new thresholds are at least as stringent as the old ones
        for key in reject:
            if key in old_reject and reject[key] > old_reject[key]:
                raise ValueError(
                    bad_msg.format(
                        kind="reject",
                        key=key,
                        new=reject[key],
                        old=old_reject[key],
                        op=">",
                    )
                )

        # same for flat thresholds
        for key in set(old_flat) - set(flat):
            flat[key] = old_flat[key]
        for key in flat:
            if key in old_flat and flat[key] < old_flat[key]:
                raise ValueError(
                    bad_msg.format(
                        kind="flat", key=key, new=flat[key], old=old_flat[key], op="<"
                    )
                )

withmywoessner replied:
One more thing, the `_reject_setup` function in epochs.py includes a section to "make sure new thresholds are at least as stringent as the old ones"

withmywoessner commented on mne/epochs.py:
```diff
@@ -818,9 +818,14 @@ def _reject_setup(self, reject, flat):
+                        % (kind, val)
                     )
 
         # now check to see if our rejection and flat are getting more
```
One more thing, the `_reject_setup` function in epochs.py includes a section to "make sure new thresholds are at least as stringent as the old ones. I think this is used if you didn't preload and you run `drop_bad()`. The current implementation just checks to see if `reject[key] > old_reject[key]`. 

This doesn't work with callables, maybe I could try checking to see if the new reject parameters would rejects less epochs than the old ones, but I would probably have to loop through all the epochs and that seems unnecessarily complex so maybe I should just not allow changing reject and flat if either is a callable.

larsoner replied:
>  I am not really sure why this conditional is necessary.

If you reject once with strict parameters A, and then ask to reject with more lenient parameters B it's a no-op and thus almost certainly an error. Users might do this thinking they'll "get back" the old ones we dropped, but once they're dropped they're gone forever.

> (My current implementation just skips the check for callables)

I think this is fine, users doing custom stuff will hopefully know once epochs are dropped they're gone.

withmywoessner commented:
@larsoner I think I am almost done with this. When you have a chance can you look over my comments and let me how you think I should address those problems. Thanks so much!

larsoner commented on mne/epochs.py:
```diff
@@ -819,9 +819,14 @@ def _reject_setup(self, reject, flat):
         for rej, kind in zip((reject, flat), ("Rejection", "Flat")):
             for key, val in rej.items():
-                if val is None or val < 0:
+                if callable(val):
```
I would simplify to:
```
if callable(val):
    continue
else:
    name = f"{kind} dict value for {key}"
    _validate_type(val, "numeric", name, extra="or callable")
    if val < 0:
         raise ValueError("{kind} {name} must be >= 0, got {val}")
```

withmywoessner replied:
I was going to do it this way, but it breaks a test. I think its because `_validate_type `returns a TypeError and Pytest wants a ValueError:
```
   for val in (None, -1):  # protect against older MNE-C types
        for kwarg in ("reject", "flat"):
            pytest.raises(
                ValueError,
                Epochs,
                raw,
                events,
                event_id,
                tmin,
                tmax,
                picks=picks_meg,
                preload=False,
                **{kwarg: dict(grad=val)},
            )
```

larsoner replied:
We should just update the test, TypeError is a better error in the case where the type is wrong (i.e., `_validate_type` does something better than what we do in `main`)

larsoner commented on mne/epochs.py:
```diff
@@ -834,33 +839,48 @@ def _reject_setup(self, reject, flat):
-                        old=old_reject[key],
-                        op=">",
+
+        if not is_callable:
+            # make sure new thresholds are at least as stringent
+            # as the old ones
```
... but I think you can make these checks more fine-grained without much more work. Instead of if *any* are callable, skip *all* checks (which is what you're doing now), you should within this loop be able to fairly easily make it so that if *this* type is or was callable, skip *this* check. I think it almost amounts to moving the conditional from above into this loop with a `continue` when callables are or were present. As a bonus, your `diff` will probably be smaller, too!

withmywoessner replied:
Should I edit my test to check for this?

larsoner commented on mne/epochs.py:
```diff
@@ -3621,25 +3641,35 @@ def _is_good(
 ):
     """Test if data segment e is good according to reject and flat.
 
+    The reject and flat dictionaries can accept functions as values.
```
```suggestion
    The reject and flat parameters can accept functions as values.
```

larsoner commented on mne/utils/docs.py:
```diff
@@ -3811,10 +3815,21 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
-              calculated for each channel separately, applying baseline
-              correction does not affect the rejection procedure, as the
-              difference will be preserved.
+    Custom rejection criteria can be also be used by passing a callable
+    to the dictionary.
+
+    Example::
+
+        reject = dict(eeg=lambda x: True if (np.max(x, axis=1) >
+                      1e-3).any() else False))
```
Simplify (no need for True if x then False... that's the same as `bool(x)` which NumPy gives you for free anyway), make it an example someone might want to use, make it a single line to render better in Sphinx, and don't word as "as the dictionary":
```suggestion
    Custom rejection criteria can be also be used by passing a callable,
    e.g., to check for 99th percentile of absolute values of any channel
    across time being bigger than 1mV::

        reject = dict(eeg=lambda x: (np.percentile(np.abs(x), 99, axis=1) > 1e-3).any())
```
You might need to add a `# noqa: E501` at the end of the `"""` for this docstring

larsoner commented on tutorials/preprocessing/20_rejecting_bad_data.py:
```diff
@@ -328,6 +330,99 @@
+    tmin=0,
+    tmax=1,
+    baseline=None,
+    reject=dict(eeg=lambda x: True if (np.max(x, axis=1) > 1e-2).any() else False),
```
```suggestion
    reject=dict(eeg=lambda x: (np.max(x, axis=1) > 1e-2).any()),
```

larsoner commented on tutorials/preprocessing/20_rejecting_bad_data.py:
```diff
@@ -328,6 +330,99 @@
+    tmin=0,
+    tmax=1,
+    baseline=None,
+    reject=dict(eeg=lambda x: True if (np.median(x, axis=1) > 1e-4).any() else False),
```
```suggestion
    reject=dict(eeg=lambda x: (np.median(x, axis=1) > 1e-4).any()),
```

larsoner reviewed (commented):
Looks like good progress so far!

> Should I add a new reason, 'REJECT_CRITERIA', for when you drop based on reject or flat that includes the function or value used to drop?

I completely forgot about this. Based on the docs of [epochs.drop(..., reason="USER")](https://mne.tools/stable/generated/mne.Epochs.html#mne.Epochs.drop), the string can actually be arbitrary. Moreover it can be a list/tuple of reasons. So how about if we require that the callable return a `good, reason` tuple like:
```
epochs.drop_bad(reject=lambda epoch: (epoch.max() > 1e-3, "> 1mV somewhere"))
```
Where `good` must be `bool` and `reason` must be `str, list, tuple` where each entry is a `str`. This allows for nice stuff like:
```
def my_reject(epoch_data):
    bad_idxs = np.where(np.percentile(epoch_data, axis=1) > 1e-3)
    reasons = tuple(epochs.ch_name[bad_idx] for bad_idx in bad_idxs)
    return len(bad_chs) > 0, reasons
```

And it also makes me think we should:

1. Update `epochs.drop(...)` to allow `reason=...` to be a *list or tuple* of reasons rather than a single string at some point. But doesn't have to be this PR unless you're really motivated.
2. Amend the `epochs.drop_log` attribute doc to mention that the entry can actually be some `ch_name` from `epochs.ch_names` as well when dropping based on flat or reject. This would be great and not too difficult to add to this PR, but not 100% necessary (I can create a new issue).

withmywoessner commented:
> 1. Update `epochs.drop(...)` to allow `reason=...` to be a _list or tuple_ of reasons rather than a single string at some point. But doesn't have to be this PR unless you're really motivated.

From my testing it seems that you can pass a tuple to reasons:
```
import mne
import matplotlib.pyplot as plt
import numpy as np
import os

# sample_data_folder = mne.datasets.sample.data_path(download=True)
sample_data_raw_file = (
    "/home/woess/mne_data/MNE-sample-data/MEG/sample/sample_audvis_filt-0-40_raw.fif"
)
raw = mne.io.read_raw_fif(sample_data_raw_file)


raw.crop(0, 5)
raw.del_proj()
chans = raw.info["ch_names"][-5:-1]
raw.pick(chans)
data = raw.get_data()

new_data = data
new_data[0, 180:200] *= 1e3
new_data[0, 460:580] += 1e-3
edit_raw = mne.io.RawArray(new_data, raw.info)

# Create fixed length epochs of 1 second
events = mne.make_fixed_length_events(edit_raw, id=1, duration=1.0, start=0)

epochs = mne.Epochs(
    edit_raw,
    events,
    tmin=0,
    tmax=1,
    baseline=None,
    #reject=dict(eeg=lambda x: True if (np.max(x, axis=1) > 1e-3).any() else False),
    preload=True,
)
epochs.drop([0,2,3], ['bad', 'bad', 'bad'])
epochs.drop_log

```
<img width="163" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/4fe44c52-4427-4b96-a20e-e2fea4bb7a7c">

> 2. Amend the `epochs.drop_log` attribute doc to mention that the entry can actually be some `ch_name` from `epochs.ch_names` as well when dropping based on flat or reject. This would be great and not too difficult to add to this PR, but not 100% necessary (I can create a new issue).

Sorry, I am a little confused, if I include this in the documentation does that mean it is not required for the callable return a `good, reason` tuple. Because if reason was required wouldn't that mean that the channel names that are currently outputted to the drop_log would always be replaced?

larsoner commented:
> From my testing it seems that you can pass a collection to reasons so I just updated the documentation
> ...
> epochs.drop([0,2,3], ['bad', 'bad', 'bad'])

Sorry, what I actually mean is that *each reason* can potentially be a tuple, e.g.:

````
epochs.drop([0, 2, 3], [('bad', 'MEG0111'), 'bad_too', ('a', 'b', 'c')])
```

> Sorry, I am a little confused, if I include this in the documentation does that mean it is not required for the callable return a good, reason tuple.

No I think the callable needs to return a `good, reason` tuple, and `reason` itself could also be a tuple of multiple reasons (or a `str` to mean one reason, equivalent to the one-reason tuple `(<str>,)`)

withmywoessner commented:
> Sorry, what I actually mean is that _each reason_ can potentially be a tuple, e.g.:

Ah okay, I think I understand my confusion. 
So basically instead of `epochs.drop([0, 2, 3], [('bad', 'a', 'b'), 'bad', ('bad', 'a')])` returning:
```
(([('bad', 'a', 'b'), 'bad', ('bad', 'a')],),
 (),
 ([('bad', 'a', 'b'), 'bad', ('bad', 'a')],),
 ([('bad', 'a', 'b'), 'bad', ('bad', 'a')],),
 ())
```
It would return:
```
(('bad', 'a', 'b'),
 (),
 'bad',
 ('bad', 'a')),
 ())
```

Got it!

withmywoessner commented:
Hey @larsoner! I haven't looked at this in a while, but I just updated the code so that drop behaves how you specified:
<img width="384" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/fb2209fb-3f55-456b-8d13-37d79d064fae">
I also added conditionals to check to see if each element in the reject and flat dicts were a tuple/list(<callable>, <list, tuple, str>) 

I have another question, which reason tuple should I use in the case of an epoch fulfilling the criteria of both reject and flat? Should I just use the reasons listed by one? Or combine them someway into a new tuple?

larsoner commented:
At the end of the day the reason needs to be a tuple of str, so if it satisfies both, you can just add the two tuples of str to get a single tuple of str. Looks like recent changes broke a lot of tests FYI

larsoner commented on mne/epochs.py:
```diff
@@ -3669,38 +3693,70 @@ def _is_good(
+                        # (bool, reason). Reason must be a str/list/tuple.
+                        # If using tuple
+                        if callable(refl):
+                            if isinstance(reasons, (tuple, list)):
+                                for idx, reason in enumerate(reasons):
+                                    _validate_type(reason, str, reason, "str")
+                                bad_tuple += tuple(reasons)
+                            if isinstance(reasons, str):
+                                bad_tuple += (reasons,)
```
A bit shorter. Adds a redundant check but fewer lines is generally nice even if it costs a tiny bit of recomputation. Also note the change in args to `_validate_type`
```suggestion
                            if isinstance(reasons, str):
                                reasons = (reasons,)
                            for idx, reason in enumerate(reasons):
                                _validate_type(reason, str, f"reasons[{idx}]")
                            bad_tuple += tuple(reasons)
```

larsoner commented on mne/epochs.py:
```diff
@@ -814,13 +814,19 @@ def _reject_setup(self, reject, flat):
+                        continue
+                    _validate_type(val, "numeric", name, extra="or callable")
+                    if val is None or val < 0:
+                        raise ValueError(
+                            """If using numerical %s criteria, the value
+                                must be >= 0 Not '%s'."""
+                            % (kind, val)
+                        )
```
Nowadays we prefer f-strings so might as well modernize since the lines are in the diff anyway. Also the triple-quoted string change you made would lead to a ton of spaces in the output error message
```suggestion
                        raise ValueError(
                            f"If using numerical {name} criteria, the value "
                            f"must be >= 0, not {repr(val)}"
                        )
```

larsoner commented on mne/epochs.py:
```diff
@@ -1522,7 +1533,7 @@ def drop(self, indices, reason="USER", verbose=None):
             correspondingly modified.
-        reason : str
+        reason : list | tuple | str
             Reason for dropping the epochs ('ECG', 'timeout', 'blink' etc).
```
```suggestion
            Reason(s) for dropping the epochs ('ECG', 'timeout', 'blink' etc).
```

larsoner commented on mne/epochs.py:
```diff
@@ -1738,7 +1758,7 @@ def _get_data(
                 if not is_good:
                     assert isinstance(bad_tuple, tuple)
-                    assert all(isinstance(x, str) for x in bad_tuple)
+                    assert all(isinstance(x, (str, tuple)) for x in bad_tuple)
```
This part is weird. For any one epoch, the `bad_tuple` should be `tuple of str`, it shouldn't be allowed to be `tuple of tuple of str`.
```suggestion
                    assert all(isinstance(x, str)) for x in bad_tuple)
```
If an epoch is rejected for example because flat check returns `tuple of str` and reject returns `tuple of str` these should be concatenated (with `+` of the two tuples). In theory you should be able to do this with empty tuples, too. I'm surprised it's not already done...?

larsoner commented on mne/epochs.py:
```diff
@@ -3669,38 +3693,70 @@ def _is_good(
+                    # Check if criterion is a function and apply it
+                    if callable(criterion):
+                        result = criterion(e_idx)
+                        _validate_type(result, tuple, result, "tuple")
```
```suggestion
                        _validate_type(result, tuple, "reject/flat output")
```

larsoner commented on mne/epochs.py:
```diff
@@ -3669,38 +3693,70 @@ def _is_good(
+                        _validate_type(result, tuple, result, "tuple")
+                        if len(result) != 2:
+                            raise TypeError(
+                                "Function criterion must return a " "tuple of length 2"
```
```suggestion
                                "Function criterion must return a tuple of length 2"
```

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -488,10 +488,11 @@ def test_average_movements():
 def _assert_drop_log_types(drop_log):
     __tracebackhide__ = True
-    assert isinstance(drop_log, tuple), "drop_log should be tuple"
+    assert isinstance(drop_log, (tuple, list)), """drop_log should be tuple
```
This change we don't want -- `drop_log` should always be immutable, so it must be a tuple (of tuple of str), not list (of tuple of str).

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -488,10 +488,11 @@ def test_average_movements():
     assert all(
-        isinstance(log, tuple) for log in drop_log
-    ), "drop_log[ii] should be tuple"
+        isinstance(log, (tuple, list)) for log in drop_log
```
Same

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -2148,6 +2174,66 @@ def test_reject_epochs(tmp_path):
+        preload=True,
+    )
+    print(epochs.drop_log)
+    assert epochs.drop_log[0] != () and epochs.drop_log[2] != ()
```
I would assert something more specific like `assert epochs.drop_log[0] == ('eeg max or median',)` or whatever it should be

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -2148,6 +2174,66 @@ def test_reject_epochs(tmp_path):
+    def reject_criteria(x):
+        max_condition = np.max(x, axis=1) > 1e-2
+        median_condition = np.median(x, axis=1) > 1e-4
+        return ((max_condition.any() or median_condition.any()), "eeg max or median")
```
Returns with commas are always implicitly tuples so this is more standard
```suggestion
        return (max_condition.any() or median_condition.any()), "eeg max or median"
```

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -3201,6 +3287,17 @@ def test_drop_epochs():
+    selection_tuple = epochs_tuple.selection.copy()
+    epochs_tuple.drop((2, 3, 4), reason=([["list"], "string", ("tuple",)]))
+    n_events = len(epochs.events)
+    assert_equal(
+        [epochs_tuple.drop_log[k] for k in selection_tuple[[2, 3, 4]]],
+        [["list"], ["string"], ["tuple"]],
+    )
```
For native Python types you should be able to use pytest directly (it overwrites `assert` to be informative on error
```suggestion
    assert [epochs_tuple.drop_log[k] for k in selection_tuple[[2, 3, 4]]] == [("list",), ("string",), ("tuple",)]
```
Note that this is also a tiny bit more specific because it asserts that the entries are tuples, whereas `assert_equal` will treat list and tuple the same

larsoner commented on mne/utils/mixin.py:
```diff
@@ -209,8 +209,23 @@ def _getitem(
             if reason is not None:
-                for k in np.setdiff1d(inst.selection, key_selection):
-                    drop_log[k] = (reason,)
+                # Used for multiple reasons
+                if isinstance(reason, tuple):
+                    reason = list(reason)
+
+                if isinstance(reason, list):
+                    for i, idx in enumerate(
+                        np.setdiff1d(inst.selection, key_selection)
+                    ):
+                        r = reason[i]
+                        if isinstance(r, str):
+                            r = (r,)
+                        if isinstance(r, list):
+                            r = tuple(r)
+                        drop_log[idx] = r
+                else:
+                    for idx in np.setdiff1d(inst.selection, key_selection):
+                        drop_log[idx] = (reason,)
```
Docstring needs to be updated above as well to mention that `reason` can be `list or tuple of str` as well. I don't think we actually want to support `reason` being a `tuple of tuple of str`, like the logic here seems to indicate. Let's just support `reason` being a `list or tuple of str` for which *all unselected epochs get the same `reason`*, i.e.,:
```suggestion
                _validate_type(reason, (list, tuple, str), "reason")
                if isinstance(reason, str):
                    reason = (reason,)
                reason = tuple(reason)
                for i, idx in enumerate(
                    np.setdiff1d(inst.selection, key_selection)
                ):
                    drop_log[idx] = reason
```

larsoner reviewed (commented):
@withmywoessner do my suggestions make sense in terms of when `tuple of str` vs `tuple of tuple of str` are used etc.?

withmywoessner commented:
> @withmywoessner do my suggestions make sense in terms of when `tuple of str` vs `tuple of tuple of str` are used etc.?

I believe so @larsoner  , It should work this, right?
![image](https://github.com/mne-tools/mne-python/assets/20428109/991e0ac7-7222-4a29-a57e-812d1c92da74)

I didn't implement that behavior for dropping using epochs constructor, just drop(). I will fix that now.

larsoner commented:
>  It should work this, right?

I was thinking no, because you can't tell whether `.drop([0, 2, 3], ('bad_a', 'bad_b', 'bad_c'))` means "use three reasons for each of 0, 2, and 3" or "use one reason for each of 0, 2, 3, corresponding to the three entries in the tuple).

I guess one way out of that is make it so that if a list or tuple is provided, it *must* be a list or tuple of list or tuple of str so the mapping is always explictly a list (one entry for each dropped index) of list of reasons.

withmywoessner commented:
> > It should work this, right?
> 
> I was thinking no, because you can't tell whether `.drop([0, 2, 3], ('bad_a', 'bad_b', 'bad_c'))` means "use three reasons for each of 0, 2, and 3" or "use one reason for each of 0, 2, 3, corresponding to the three entries in the tuple).
> 
> I guess one way out of that is make it so that if a list or tuple is provided, it _must_ be a list or tuple of list or tuple of str so the mapping is always explictly a list (one entry for each dropped index) of list of reasons.

The way I intended it was that if you pass just a string it applies that string to every drop_log entry like this:
<img width="446" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/a186684b-5251-4e9f-9a72-959207b3ad67">

But if you pass a tuple or a list, it does what I specified above. 

I'll do what you think is best though!

larsoner commented:
Yes I think single `str` broadcasting to `tuple of tuple of that one str` is fine, then supporting *only* `tuple-like of tuple-like of str` and *not* `tuple-like of str` (as this case can be ambiguous) seems okay!

And at the end of the day `epochs.drop_log` should *always* be `tuple of tuple of str`

withmywoessner commented:
> Yes I think single `str` broadcasting to `tuple of tuple of that one str` is fine, then supporting _only_ `tuple-like of tuple-like of str` and _not_ `tuple-like of str` (as this case can be ambiguous) seems okay!
> 
> And at the end of the day `epochs.drop_log` should _always_ be `tuple of tuple of str`

I may be fixating one this too

> Yes I think single `str` broadcasting to `tuple of tuple of that one str` is fine, then supporting _only_ `tuple-like of tuple-like of str` and _not_ `tuple-like of str` (as this case can be ambiguous) seems okay!
> 
> And at the end of the day `epochs.drop_log` should _always_ be `tuple of tuple of str`

Maybe its just simpler to only allow a tuple of str or str? And not allow tuple of tuple of string. I think that matches the original functionality of reason better which currently only allows str, but I'll leave it alone if you think otherwise.

larsoner commented:
> Maybe its just simpler to only allow a tuple of str or str? And not allow tuple of tuple of string. I think that matches the original functionality of reason better which currently only allows str, but I'll leave it alone if you think otherwise.

This would be okay, too, if that `tuple of str` is used as the reason for all dropped indices, like:
```
epochs.drop([0, 2, 3], ('a', 'b', 'c'))
```
yields
```
epochs.drop_log == (('a', 'b', 'c'), (), ('a', 'b', 'c'), ('a', 'b', 'c'))
```

withmywoessner commented:
> Yields
> 
> ```
> epochs.drop_log == (('a', 'b', 'c'), (), ('a', 'b', 'c'), ('a', 'b', 'c'))
> ```
Is this okay?
`epochs.drop([0, 2, 3], ('a', 'b', 'c'))
`
Yields
`epochs.drop_log == (('a',), (), ('b',), ('c',))`

withmywoessner commented:
> I don't like that one because currently the way it works is you give one reason, it's applied to all indices. In your suggestion, you are allowed to give a different single reasons for different indices. But there is no way to give multiple reasons for any indices.

Maybe I just don't understand the benefit of giving multiple reasons as a tuple. I would assume that if someone is manually specifying something they would want it to be considered one reason. 

I just wanted to include something that allowed you to give a different reason for each epoch, but I guess you can just call drop multiple times anyway. 

I think I have wasted too much time on this so I will just defer to you. Thanks for the clarification!!

larsoner commented:
> Maybe I just don't understand the benefit of giving multiple reasons as a tuple. I would assume that if someone is manually specifying something they would want it to be considered one reason.

I'm thinking of something like `('bad_ch_1', 'bad_ch_2', 'bad_blink')` or whatever being useful in the drop log, just like having a list of channel names from reject/flat could be.

> I just wanted to include something that allowed you to give a different reason for each epoch, but I guess you can just call drop multiple times anyway.

Yes exactly! I think this case is easily done within a loop just like it has to be in `main` currently. Hence the `tuple of str` meaning "use this tuple of reasons for all indices" extends functionality in a way there is no workaround/other workflow for currently I think.

larsoner commented:
... although I guess we could do something like what you propose if we say:

```
reason : str | list or tuple of str | list or tuple of list or tuple of str
    The reason for drops. If list or tuple, must have same length as `drops`,
    and each entry will be the reason (or list/tuple of reasons) for each
    index's drops
```
then I guess the case of 
```
epochs.drop([0, 1, 2], ('a', 'b', 'c'))
```
*will* behave like you suggest, and is unambiguous at least from a docstring perspective. And the behavior I want can easily be achieved with
```
epochs.drop([0, 1, 2], [('a', 'b', 'c')] * 3)
```
So if you want to go this direction intead, that would work! And apologies if this is what you already have, I should have thought this through a bit further probably 😬

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -488,10 +488,11 @@ def test_average_movements():
 def _assert_drop_log_types(drop_log):
     __tracebackhide__ = True
-    assert isinstance(drop_log, tuple), "drop_log should be tuple"
+    assert isinstance(drop_log, tuple), """drop_log should be tuple
+    or list"""
```
```suggestion
    assert isinstance(drop_log, tuple), "drop_log should be tuple"
```

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -488,10 +488,11 @@ def test_average_movements():
     assert all(
         isinstance(log, tuple) for log in drop_log
-    ), "drop_log[ii] should be tuple"
+    ), "drop_log[ii] should be tuple or list"
```
```suggestion
    ), "drop_log[ii] should be tuple"
```

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -563,6 +564,31 @@ def test_reject():
+    bad_types = [my_reject_1, my_reject_2, ("Hi" "Hi"), (1, 1), None]
+    for val in bad_types:  # protect against bad types
+        for kwarg in ("reject", "flat"):
+            pytest.raises(
```
This is the old style, better is:
```
with pytest.raises(TypeError, match="..."):
    Epochs(...)
```

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -3181,7 +3282,9 @@ def test_drop_epochs():
     pytest.raises(IndexError, epochs.drop, [-len(epochs.events) - 1])
-    pytest.raises(ValueError, epochs.drop, [[1, 2], [3, 4]])
+    pytest.raises(TypeError, epochs.drop, [[1, 2], [3, 4]])
+    with pytest.raises(TypeError):
```
should have `match="..."` to both make it clear what's being caught, and make sure that's what is actually being caught

larsoner commented on mne/tests/test_epochs.py:
```diff
@@ -3181,7 +3282,9 @@ def test_drop_epochs():
     pytest.raises(IndexError, epochs.drop, [len(epochs.events)])
     pytest.raises(IndexError, epochs.drop, [-len(epochs.events) - 1])
-    pytest.raises(ValueError, epochs.drop, [[1, 2], [3, 4]])
+    pytest.raises(TypeError, epochs.drop, [[1, 2], [3, 4]])
```
Since you're touching this one anyway, can you modernize it with `with pytest.raises(..., match=...)` style? Feel free to update the others if you're up for it (or not!)

larsoner reviewed (commented):
Very close, just a few remaining minor things! Can you also add a `doc/changes/devel/12195.newfeature.rst` and update to latest `main` (e.g., with `git fetch upstream` then `git merge upstream/main`)? It will bring in some CI changes to build requirements. (Or you or I can click the button on GitHub to update to latest main, too.)

withmywoessner commented:
Done! Thanks so much for all your help! @larsoner Maybe this can be another PR, but one thing I realized is that I didn't include something to check if the callable returns a valid properly formatted tuple if `preload=False`. I don't think this is possible as it requires me to actually call it. I tried including a warning when you pass a callable and `preload = False`, but for whatever reason I couldn't get the proper truth value of `preload`. Do you think this is something I need to add in this or a seperate PR? Again thanks so much for the help!

larsoner commented:
> Maybe this can be another PR, but one thing I realized is that I didn't include something to check if the callable returns a valid properly formatted tuple if preload=False

Ohh, because it's not called right away, right? This actually raises a potential issue -- we don't want `epochs.reject` to ever be anything other than a `dict` (mapping str keys to *float vals*) or `None`, and it will be after this if you allow these to be passed in the `Epochs` constructor. This will make it so we can't write `epochs` to disk anymore. See for example in your tutorial if I add a `epochs.save('test-epo.fif')`:
```
Traceback (most recent call last):
  File "/home/larsoner/python/mne-python/tutorials/preprocessing/20_rejecting_bad_data.py", line 383, in <module>
    epochs.save("test-epo.fif")
  File "<decorator-gen-199>", line 12, in save
  File "/home/larsoner/python/mne-python/mne/epochs.py", line 2226, in save
    over_size += len(json.dumps(reject_params)) + 16
                     ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type function is not JSON serializable
```

To fix this problem and the output-not-checked problem, rather than supporting callables in the constructor `Epochs.__init__`, we could just support callables in `epochs.drop_bad`. In other words, this *should* work (which doesn't currently):
```
>>> import mne
>>> raw = mne.io.read_raw_fif(mne.datasets.sample.data_path() / "MEG" / "sample" / "sample_audvis_raw.fif")
>>> epochs = mne.make_fixed_length_epochs(raw)
>>> epochs.drop_bad(reject=dict(eeg=lambda x: (x.max() > 1e-5, "bad")))
```
Whereas `Epochs(..., reject=dict(eeg=lambda x: ...))` should *not*.

Then we can modify `epochs.reject` within `def drop_bad` so that saving will work. And the correctness will be checked right away, because `.drop_bad` operates immediately (does not wait until later).

To do this I think you can keep most of what you changed, but modify `def _reject_setup(reject, flat, *, allow_callable=False)` as the signature so that when it's called in `__init__` it won't allow callables, but then in `drop_bad` it calls it with `self._reject_setup(reject, flat, allow_callables=True)`, and then after the call to `self._get_data` it removes the keys in `self.reject` and `self.flat` that were callables. WDYT?

Another option I guess would be to make it so that on save we check the reject and flat and remove the callables then instead, which would probably be okay, too. Then the output correctness won't get checked right away, though.

larsoner commented:
Yes it's possible to check `preload`. But if you go the `.drop_bad(...)`-only route it's not necessary. It will go through and load each epoch from `raw` on demand (or from ndarray if it's preloaded) and decide whether to keep or reject it. So maybe it's cleaner to use that rather than try to support it in `__init__(...)` with some processing delayed until later.

Would supporting just `.drop_bad(...)` work for your use case?

withmywoessner commented:
Yes @larsoner! I think I might write new docustrings for `drop_bad` because it currently it shares docustrings with reject and flat in the constructor of `Epochs`. I think it would be confusing if this functionality was mentioned in the docustrings for both.

withmywoessner commented:
Hey @larsoner, just pinging here. I think everything should be ready finally. Thanks so much for helping me with this :)

larsoner commented on doc/changes/devel/12195.newfeature.rst:
```diff
@@ -0,0 +1 @@
+Add ability reject :class:`mne.Epochs` using callables.
```
```suggestion
Add ability reject :class:`mne.Epochs` using callables, by `Jacob Woessner`_.
```

larsoner reviewed (approved):
Just two tiny things I noticed, will commit and mark for merge-when-green. Thanks for working through my confusions on this one @withmywoessner !

drammock commented:
maybe I'm misunderstanding something, but it looks like Circle's `build_docs` job passed in 8 seconds (?!) on the last commit of this PR, and didn't actually build the docs... which are now failing on #12411 with 

```
/home/circleci/project/mne/epochs.py:docstring of mne.epochs.drop_bad:49: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```

@larsoner any idea why we got a false-positive pass from Circle here?

larsoner commented:
Looks like for some reason it ran on his fork rather than our project. Not sure why

withmywoessner commented:
> Looks like for some reason it ran on his fork rather than our project. Not sure why

https://support.circleci.com/hc/en-us/articles/360008097173-Troubleshooting-why-pull-requests-are-not-triggering-jobs-on-my-organization

Is this the reason why? @larsoner @drammock 

> A user who submits a pull request to your repository from a fork, but no pipeline is triggered with the pull request. This can happen when the user is following the project fork on their personal account rather than the project itself on CircleCI.

> In these cases, the user unfollows their fork of the project on CircleCI. This will trigger their jobs to run under the organization when they submit pull requests. Those users can optionally follow the source project if they wish to see the pipelines.

I clicked unfollow.

drammock commented:
> I clicked unfollow.

thanks! hopefully that will fix it for your next PR.

drammock commented:
I think what is supposed to happen is that if you create a CircleCI account that is linked to your GH account (maybe they always are?) then Circle will run the job automatically *as long as it's not your first PR in the MNE-Python repo*.  But to be honest I haven't really had time to look into the situation / failures very carefully.  Maybe I should create a new GH account so I can try it out / experience the rejection myself, and then document the fix 😆

drammock commented:
I don't know :(  I just poked around in the admin interface and I see no list of "members" and no way to "add" people...  your list of organizations looks odd though (mine lists various user's names with images, and various named orgs, but no hash-like codes like `cci-3rq7p` like yours does). Do you see a link at the bottom of that dropdown saying "Can't find an organization? Check permissions"

withmywoessner commented:
Yes, 
<img width="214" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/323092a9-79ab-4d42-95de-a864311e35e1">
I click that and it takes me here:
<img width="892" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/c85038fb-7c18-4dea-b5bc-b7dce1d3e05f">
I click, 'check permissions...' and it takes me:
<img width="902" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/5e0a7655-bb47-4a91-ba13-8ae3873d0d65">

Is there anything linking your CircleCI account to the mne-python public repo on your CircleCI account page? I don't see anything like that on mine.

Thanks for looking into this!

drammock commented:
ok, IIUC they are using GitHub users & organizations as their source of "organizations" and since you're not a member of https://github.com/orgs/mne-tools you shouldn't expect to see "mne-tools" as an option within CircleCI. But since I *am* a member, I do see "mne-tools" as an org within CircleCI (along with other orgs I'm part of: pydata, sphinx-gallery, etc).

So I think that part at least is working as expected? (I don't see any orgs listed on your GitHub profile so it makes sense you wouldn't see any within CircleCI).  But I think (?) this shouldn't prevent the jobs from running; hopefully it was the "follow" setting that you've just changed; we'll know on your next PR I guess.

withmywoessner commented:
Hopefully, @drammock 😬. Is there a way to change where CircleCI is run from in an existing PR?
Looking at my open PR #12392. I just merged the branch and it still says the tests are being run from my branch:
<img width="774" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/f9dc37cd-42a8-471c-9106-b0f9baa6591b">

Compared to this one from #12382:
<img width="849" alt="image" src="https://github.com/mne-tools/mne-python/assets/20428109/945aa6e0-f062-4996-8b59-06dce02db332">

It also has a different icon in the top left and says 'mne-tools' when I clicked on it.

drammock commented:
#12392 looks correct to me:

![Screenshot 2024-02-02 at 14-51-13 build_docs (63616) - mne-tools_mne-python](https://github.com/mne-tools/mne-python/assets/1810515/e5fe3603-d88c-4480-a3c2-ce983329f029)

not sure why you're seeing something different. To get to that page I clicked on the "Details" link on #12392:

![Screenshot 2024-02-02 at 14-52-40 Create functions to read Epochs and Evoked Neuroscan formats by withmywoessner · Pull Request #12392 · mne-tools_mne-python](https://github.com/mne-tools/mne-python/assets/1810515/998b8d7d-2d78-407e-8ab6-7c1f0cc8eec7)

withmywoessner commented:
It looks good now. Maybe I just clicked it too early. Thanks!
